### PR TITLE
Run all OnExit before all OnEnter

### LINKED ISSUE #10732: Run all OnExit before all OnEnter
`StateTransition` schedule executes `apply_state_transition` system for each state in order. `apply_state_transition` executes `OnExit`, `OnTransition`, `OnEnter` schedules in order.

https://github.com/bevyengine/bevy/blob/4eafd60ce9dd3dd579c07798ceb43f88a13d7094/crates/bevy_ecs/src/schedule/state.rs#L154-L161

This is somewhat inconvenient.

For example, application state can be controller by two state types, `StateA` and `StateB` (for example, state and substate).

Each state may want to for example, set cursor icon on enter and reset it to default on exit (clean up). This is not possible now, because cleanup of second system (`OnExit`) can be executed after init of first system (`OnEnter`).

Propose to execute systems in order:

```
OnExit<A>
OnExit<B>
OnTransition<A>
OnTransition<B>
OnEnter<A>
OnEnter<B>
```

### PR DESCRIPTION
**Pull request by stepancheg**
Fixes #10732. Copying description from that issue.

# Objective

`StateTransition` schedule executes `apply_state_transition` system for each state in order. `apply_state_transition` executes `OnExit`, `OnTransition`, `OnEnter` schedules in order.

https://github.com/bevyengine/bevy/blob/4eafd60ce9dd3dd579c07798ceb43f88a13d7094/crates/bevy_ecs/src/schedule/state.rs#L154-L161

This is somewhat inconvenient.

For example, application state can be controller by two state types, `StateA` and `StateB` (for example, state and substate).

Each state may want to for example, set cursor icon on enter and reset it to default on exit (clean up). This is not possible now, because cleanup of second system (`OnExit`) can be executed after init of first system (`OnEnter`).

## Solution

Execute systems in order:

```rust
OnExit<A>
OnExit<B>
OnTransition<A>
OnTransition<B>
OnEnter<A>
OnEnter<B>
```

## Changelog

- Deprecate `apply_state_transition::<T>` system
- Add `apply_state_transition_on_exit::<T>`, `apply_state_transition_on_transition::<T>`, `apply_state_transition_on_enter::<T>` systems
- Add non-generic `AfterOnExit`, `BeforeOnEnter` labels to order the systems

## Migration Guide

No migration needed if only one state was used which was initialized with `.add_state`.

If more than one state is used, order of state callback execution was changed, need to review the code to make sure it is not broken.

If `apply_state_transition` system was used directly, it will still work, but this function is deprecated and will be eventually removed.

## Side note

I proposed implicit system ordering (#10678, #10618) before, and I think this is another illustration where implicit ordering would be beneficial.

In current setup (before or after this PR) all systems are exclusive and thus not parallel. I imagine it could be like this:

```rust
app.add_systems(
  ScheduleTransition,
  enter_main_menu.on_enter(GameState::MainMenu),
)
```

~~`enter_main_menu.on_enter(GameState::MainMenu)` could create a system with after-dependency on non-generic `BeforeOnEnter`.~~ **Edit**: actually we can already do that, by adding the same before/after as this diff does.

So all the systems in `ScheduleTransition` will run concurrently, but with proper ordering, no subschedules.

Moreover, we would not even need to run `enter_main_menu` in `ScheduleTransition`, we could run it in `Update` (or any other schedule), if we could implicitly order all regular systems which read or update state after all systems which are triggered on state change.

But that is somewhat offtopic here. For now I mostly want to deal with this ordering issue.

### CODE REVIEW

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Definitely warming up to this architecture!

Rather than using `AfterOnExit` and `BeforeOnEnter` sets, we should do:

```rust
#[derive(SystemSet)
pub enum StateTransitionSet {
    OnExit,
    OnTransition,
    OnEnter
}
```
Then, add your systems to these sets, then use `configure_sets` to order these sets.

**Comment by stepancheg** [comment]
This is broken version but I don't see convert to draft button.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 1]
```diff
@@ -135,29 +135,80 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
         .ok();
 }
 
+/// Label to order all [`OnEnter`], [`OnTransition`], and [`OnExit`] systems.
+///
+/// All `OnEnter` will run before all `OnTransition`,
+/// and all `OnTransition` will run before all `OnExit`.
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/state.rs:141]:
```suggestion
/// and all `OnTransition` will run before all `OnExit`.
///
/// The order of `OnEnter` systems relative to each other (and so on) is ambiguous by default.
```

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/schedule/state.rs [hunk 2]
```diff
@@ -135,29 +135,82 @@ pub fn run_enter_schedule<S: States>(world: &mut World) {
...
 /// - Takes the new state value from [`NextState<S>`] and updates [`State<S>`].
-/// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
-/// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
 /// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
-pub fn apply_state_transition<S: States>(world: &mut World) {
+pub fn apply_state_transition_on_enter<S: States>(world: &mut World) {
     // We want to take the `NextState` resource,
     // but only mark it as changed if it wasn't empty.
+    // Note we update the `State` and `NextState` resources only in on enter,
+    // because previous two systems need both the old and new state.
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:205]: This is actually a breaking change in behavior. Previously, `apply_state_transition` would `std::mem::take` the value from `NextState`, then replace the value of `State`. Then it would call `OnExit`, `OnTransition`, `OnEnter`. All in that order. I think it makes sense to preserve this order of operations.

One way of preserving the old ordering is to store the old value from a state transition in a `PreviousState` resource, which would later be used to call `OnExit` and `OnTransition`.

```rust
fn apply_state_transition<S: States>(
    mut next_state: ResMut<NestState<S>>,
    mut state: ResMut<State<S>>,
    mut prev_state: ResMut<PreviousState<S>>,
) {
     // Swappy swap. `next_state` -> `state` -> `prev_state`.
     // Nothing else happens here.
     // `OnEnter` and friends get called elsewhere.
}
```

The `.add_systems` would look something like

```rust
.add_systems((
    apply_state_transition,
    // These SystemSets could be private.
    run_exit_schedule.in_set(RunOnEnter),
    run_transition_schedule.in_set(RunOnTransition),
    run_enter_schedule.in_set(RunOnEnter),
).chain())
.configure_sets((
    RunOnExit,
    RunOnTransition,
    RunOnEnter,
).chain());
```

Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs:205]: > apply_state_transition

```
fn apply_state_transition<S: States>(
    mut next_state: ResMut<NestState<S>>,
    mut state: ResMut<State<S>>,
    mut prev_state: ResMut<PreviousState<S>>,
) {
     // Swappy swap. `next_state` -> `state` -> `prev_state`.
     // Nothing else happens here.
     // `OnEnter` and friends get called elsewhere.
}
```

to make this work we should make this this function private or rename it to avoid break backwards compatibility too hard — if someone uses this function directly.

> store the old value from a state transition in a PreviousState

This can be done.

Except maybe this "swappy swap" should happen after `OnExit`, because while we are running `OnExit` we are logically still in the previous state.

So do you think it is important to preserve current behavior?

And more generally, is it important to guarantee specific values of `State` and `NextState` resources during state transition?

Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs:205]: > Except maybe this "swappy swap" should happen after OnExit, because while we are running OnExit we are logically still in the previous state.

I think that if we're running `OnExit(A)` when we leave the state `A`, that means we should already be in the next state.

Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs]: Restored previous behavior:
* `PrevState` now holds previous value
* `State` is updated to new state, and `NextState` is reset to `None` before all three systems run

File: crates/bevy_ecs/src/schedule/state.rs [hunk 3]
```diff
@@ -128,37 +129,95 @@ impl<S: States> NextState<S> {
...
-    // We want to take the `NextState` resource,
-    // but only mark it as changed if it wasn't empty.
+    apply_state_transition_on_enter::<S>(world);
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnExit(exited_state)`] schedule, if it exists.
+pub fn apply_state_transition_on_exit<S: States>(world: &mut World) {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:175]: In order to get the most consistent ordering, I think we should perform the state transition in a separate system before calling `OnExit`. This would counteract the stochastic execution order of the state transition systems by ensuring that every state has completed its transition by the time any of the `OnExit` systems run.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 4]
```diff
@@ -128,37 +129,95 @@ impl<S: States> NextState<S> {
     }
 }
 
+/// The previous state of [`State<S>`].
+///
+/// This is used internally by Bevy and generally not meant to be used directly.
+/// Precise semantics of data in this resource are not specified.
+#[derive(Resource, Default, Debug)]
+pub struct PrevState<S: States>(pub Option<S>);
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:137]: We can probably unset this after calling `OnTransition`

File: crates/bevy_ecs/src/schedule/state.rs (hunk 3 above)
Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs:175]: Makes sense.

File: crates/bevy_ecs/src/schedule/state.rs (hunk 4 above)
Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs:137]: Aren't you as paranoid about performance of running extra system per state 60 times per second as I am?

Probably it is nothing, but this is the only argument I have against running it as a separate system after `on_enter`. From design point of view, this would be best.

**Comment by stepancheg** [comment]
Now all the transition systems observe consistent states of every state (not just own).

File: crates/bevy_ecs/src/schedule/state.rs [hunk 5]
```diff
@@ -128,37 +129,113 @@ impl<S: States> NextState<S> {
...
-            // Try to run the schedules if they exist.
-            world.try_run_schedule(OnExit(exited.clone())).ok();
-            world
-                .try_run_schedule(OnTransition {
-                    from: exited,
-                    to: entered.clone(),
-                })
-                .ok();
-            world.try_run_schedule(OnEnter(entered)).ok();
+            world.resource_mut::<PrevState<S>>().0 = Some(exited.clone());
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:194]:
```suggestion
            world.resource_mut::<PrevState<S>>().0 = Some(exited);
```

I think this clone is unnecessary.

**Comment by JoJoJet** [comment]
If you're worried about overhead from adding all these extra systems, we can eliminate most of that by adding run conditions to the `on_exit`, `on_transition`, and `on_enter` systems, which will prevent the executor from even spawning a task for these systems unless a state transition is happening. If we attach the run condition to a system set, then it will only be evaluated once for each system type which avoids even more overhead.

**Comment by stepancheg** [comment]
@JoJoJet like this?

```rust
            .add_systems(
                StateTransition,
                (
                    apply_state_transition_before_transition::<S>
                        .in_set(StateTransitionSet::BeforeTransition),
                    apply_state_transition_on_exit::<S>.in_set(StateTransitionSet::OnExit),
                    apply_state_transition_on_transition::<S>
                        .in_set(StateTransitionSet::OnTransition),
                    apply_state_transition_on_enter::<S>.in_set(StateTransitionSet::OnEnter),
                    apply_state_transition_after_transition::<S>
                        .in_set(StateTransitionSet::AfterTransition),
                // HERE
                ).run_if(next_state_was_updated::<T>),
            );
```

Should I do it in this PR or in follow-up?

**Comment by JoJoJet** [comment]
That's a lot simpler than what I was picturing, but yeah that should work perfectly. Just make it check if `Res<NextState>` is `Some`

File: crates/bevy_ecs/src/schedule/state.rs [hunk 6]
```diff
@@ -128,37 +129,113 @@ impl<S: States> NextState<S> {
...
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+    apply_state_transition_after_transition::<S>(world);
+}
+
+/// Change the state before invoking transition systems.
+pub fn apply_state_transition_before_transition<S: States>(world: &mut World) {
+    // Reset the previous state, so the following systems which rely only on `PrevState` won't run.
+    let mut prev_state_resource = world.resource_mut::<PrevState<S>>();
+    if prev_state_resource.bypass_change_detection().0.is_some() {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:183]:
```suggestion
    if prev_state_resource.0.is_some() {
```

**Comment by JoJoJet** [comment]
Might as well make the change in this PR since it's quite small.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 7]
```diff
@@ -128,37 +129,113 @@ impl<S: States> NextState<S> {
...
-    // but only mark it as changed if it wasn't empty.
+    apply_state_transition_before_transition::<S>(world);
+    apply_state_transition_on_enter::<S>(world);
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+    apply_state_transition_after_transition::<S>(world);
+}
+
+/// Change the state before invoking transition systems.
+pub fn apply_state_transition_before_transition<S: States>(world: &mut World) {
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:180]: This doesn't need to be an exclusive system anymore. We can get easy parallelism by accessing the state resources directly.

Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs:180]: To do this we need to kill legacy `apply_state_transition` first, which calls these systems (as functions) directly (or duplicate a lot of code).

This is much change than it seems.

Suggestions? Or we can do it in follow-up PR.

Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs:180]: I believe you could use `SystemState` to call this function from an exclusive system.

```rust
fn apply_state_transition_before_transition<S: States>(
    mut prev_state: ResMut<PrevState<S>>,
    mut state: ResMut<State<S>>,
    mut next_state: ResMut<NextState<S>>,
) {
...
}

fn apply_state_transition<S: States>(
    world: &mut World,
    before_transition_state: &mut SystemState<(
        ResMut<PrevState<S>>,
        ResMut<State<S>>,
        ResMut<NextState<S>>,
    )>,
) {
     let (prev, curr, next) = before_transition_state.get(world);
     apply_state_transition_before_transition(prev, curr, next);
}
```

Or if you don't want to repeat the function parameters you could use a type alias or custom `SystemParam`

**Comment by stepancheg** [comment]
Rewritten some code:
* five systems are now guarded by a common condition
* systems made private and instead `apply_state_transition_systems` function exposes them
* each those systems now assume states are correct instead of checking state
* added state `StateTransitionSet::RunEnterOnce` to make first on-enter also safe

**Comment by stepancheg** [comment]
Removed `World` in `apply_before_transition` and `apply_after_transition`.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 8]
```diff
@@ -128,37 +127,166 @@ impl<S: States> NextState<S> {
...
+    after_transition_states: &mut SystemState<ResMut<PrevState<S>>>,
+) {
+    let state_changed = {
+        let states = next_state_changed_states.get(world);
+        next_state_changed(states.0, states.1)
+    };
+    if state_changed {
+        {
+            let states = before_transition_states.get_mut(world);
+            apply_before_transition::<S>(states.0, states.1, states.2);
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:188]:
```suggestion
            let (prev, curr, next) = before_transition_states.get_mut(world);
            apply_before_transition::<S>(prev, curr, next);
```

File: crates/bevy_ecs/src/schedule/state.rs [hunk 9]
```diff
@@ -128,37 +127,166 @@ impl<S: States> NextState<S> {
...
+    before_transition_states: &mut SystemState<(
+        ResMut<PrevState<S>>,
+        ResMut<State<S>>,
+        ResMut<NextState<S>>,
+    )>,
+    after_transition_states: &mut SystemState<ResMut<PrevState<S>>>,
+) {
+    let state_changed = {
+        let states = next_state_changed_states.get(world);
+        next_state_changed(states.0, states.1)
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:183]:
```suggestion
        let (state, next_state) = next_state_changed_states.get(world);
        next_state_changed(state, next_state)
```

File: crates/bevy_ecs/src/schedule/state.rs [hunk 10]
```diff
@@ -128,37 +127,166 @@ impl<S: States> NextState<S> {
...
+            .in_set(StateTransitionSet::RunEnterOnce),
+        (
+            apply_before_transition::<S>.in_set(StateTransitionSet::BeforeTransition),
+            run_on_exit::<S>.in_set(StateTransitionSet::OnExit),
+            run_on_transition::<S>.in_set(StateTransitionSet::OnTransition),
+            run_on_enter::<S>.in_set(StateTransitionSet::OnEnter),
+            apply_after_transition::<S>.in_set(StateTransitionSet::AfterTransition),
+        )
+            .run_if(next_state_changed::<S>),
+    ))
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:222]:
```suggestion
    (
        run_enter_schedule::<S>
            .run_if(common_conditions::run_once())
            .in_set(StateTransitionSet::RunEnterOnce),
        (
            apply_before_transition::<S>.in_set(StateTransitionSet::BeforeTransition),
            run_on_exit::<S>.in_set(StateTransitionSet::OnExit),
            run_on_transition::<S>.in_set(StateTransitionSet::OnTransition),
            run_on_enter::<S>.in_set(StateTransitionSet::OnEnter),
            apply_after_transition::<S>.in_set(StateTransitionSet::AfterTransition),
        )
            .run_if(next_state_changed::<S>),
    )
        .chain()
```

Even though `App::add_state` adds orderings to their system sets, we should make sure that standalone usage of `apply_state_transition_systems` runs these in the right order.

Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs]: Done.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 11]
```diff
@@ -128,37 +127,166 @@ impl<S: States> NextState<S> {
...
+
+    // Try to run the schedules if they exist.
+    world.try_run_schedule(OnEnter(entered)).ok();
+}
+
+/// Change the state after invoking transition systems.
+fn apply_after_transition<S: States>(mut prev: ResMut<PrevState<S>>) {
+    let prev = prev.0.take();
+
+    assert!(
```
[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs]: I suggested changing to a `debug_assert` since I don't want to cause production builds to crash over something like this. I wonder if we could just make `PrevState` private, then there would be basically no chance of this assertion failing.

[the remark this answers is not in the document]
Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs]: >  I don't want to cause production builds to crash over something like this

In my opinion it is better to crash at production early than have some other problem in production (including crash in other placer) which is much harder to debug (may be orders of magnitude harder) than assertion (for example, corrupted save files).

> I wonder if we could just make PrevState private, then there would be basically no chance of this assertion failing.

Locally in my project I have `InitResMut`, which is a `ResMut` wrapper which also does `insert_resource` in `init_state`. I think this goes against Bevy model. But it is possible.

Another possibility is mark `PrevState` `#[doc(hidden)]`.

Another possibility is to hide resource initialization in a function like `init_resources_for_state<S: States>(world: &mut World)`.

[the remark this answers is not in the document]
Comment (JoJoJet) [reply crates/bevy_ecs/src/schedule/state.rs]: > In my opinion it is better to crash at production early than have some other problem in production (including crash in other placer) which is much harder to debug (may be orders of magnitude harder) than assertion (for example, corrupted save files).

The way I see it, the goal of a production build is not to be debuggable. It's to provide a reliable and smooth user experience. In release builds, it's sometimes better to keep chugging along in an invalid state rather than crashing the whole application, especially when the invalid state is as minor as this one.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 12]
```diff
@@ -128,37 +127,172 @@ impl<S: States> NextState<S> {
...
+/// And on state change, the following schedules will be run in order:
+/// * [`OnExit<S>`]
+/// * [`OnTransition<S>`]
+/// * [`OnEnter<S>`]
+///
+/// This is called by `App::add_state`, and is not meant to be used directly.
+pub fn apply_state_transition_systems<S: States>() -> SystemConfigs {
+    // Extra `.chain()` calls case `apply_state_transition_systems` is used directly
+    // without properly configuring the `StateTransitionSet` ordering.
+    IntoSystemConfigs::into_configs(
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:212]: I think this call to `IntoSystemConfigs::Into_configs` is unnecessary, since calling `.chain()` on the tuple returns a `SystemConfigs` instance

File: crates/bevy_ecs/src/schedule/state.rs [hunk 13]
```diff
@@ -128,37 +127,172 @@ impl<S: States> NextState<S> {
...
+        (
+            run_enter_schedule::<S>
+                .run_if(common_conditions::run_once())
+                .in_set(StateTransitionSet::RunEnterOnce),
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:216]: I'm thinking maybe we shouldn't include the `run_enter_schedule.run_if(run_once())` system in `apply_state_transition_systems`. I would like it if you could add `apply_state_transition_systems<T>` to multiple schedules in order to allow multiple state transitions per frame, but the system that runs once kind of breaks that.

We should probably not include this system in then function, and have `App::add_state` add it separately.

File: crates/bevy_ecs/src/schedule/state.rs [hunk 14]
```diff
@@ -128,37 +127,172 @@ impl<S: States> NextState<S> {
...
+}
+
+/// If a new state is queued in [`NextState<S>`], this system
+/// runs the [`OnEnter(entered_state)`] schedule, if it exists.
+fn run_on_enter<S: States>(world: &mut World) {
+    let entered = world.resource::<State<S>>().0.clone();
+
+    // Try to run the schedules if they exist.
+    world.try_run_schedule(OnEnter(entered)).ok();
+}
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/state.rs:288]: Can we combine this with the `run_enter_schedule` function?

File: crates/bevy_ecs/src/schedule/state.rs (hunk 13 above)
Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs]: OK.

**Comment by JoJoJet** [comment]
Thank you!

File: crates/bevy_ecs/src/schedule/state.rs [hunk 15]
```diff
@@ -128,37 +127,155 @@ impl<S: States> NextState<S> {
...
+            let states = after_transition_states.get_mut(world);
+            apply_after_transition::<S>(states);
+        }
+    }
+}
+
+/// Configure systems running state transition callbacks.
+///
+/// [`OnEnter<S>`] schedule will be run once on first run.
+/// And on state change, the following schedules will be run in order:
```
[the remark this answers is not in the document]
Comment (stepancheg) [reply crates/bevy_ecs/src/schedule/state.rs]: Thanks!

**Comment by stepancheg** [comment]
Is anything else needed to be done to merge this PR?

Making code work correctly with multiple states is tedious for scenarios I explained: set up something when entering state, and cleanup when leaving state need to be done very carefully when state enter/exit are executed in unknown order.

**Comment by JoJoJet** [comment]
Previously we were waiting on two approvals from SMEs, but I suppose we have  those now. @alice-i-cecile,  think we should merge this?

**Comment by alice-i-cecile** [comment]
@maniwani @hymm I'd like a last set of opinions from at least one of you. I'm nervous there's something I'm missing. The loss of atomicity is annoying, but I think that the benefits are worth it.

**Comment by hymm** [comment]
I'm ambivalent about this PR. It does make things a little nicer by making the ordering predictable, but what we really need is a solution for sub states and this doesn't feel like it gets us any closer to that. This has an incorrect ordering for sub states. i.e. If `B` is a substate of `A` the correct ordering ignoring the transition schedules is `OnEnter<A>`, `OnEnter<B>`, `OnExit<B>`, `OnExit<A>`. This is so all the B state is cleaned up before trying to exit the A state. So when states have a dependency on each other they have an inside outside relationship with each other. The ordering described in this pr is just arbitrary. But arbitrary is probably better than random. But this is a lot of complexity over the previous code, so I could go either way.

One question I have is what would adding a state to the render schedule look like after this PR?

**Comment by stepancheg** [comment]
> what we really need is a solution for sub states

Also it can actually built on top of that. Relatively simple API, and more or less compatible with current, might be:

```
trait States {
  // New member.
  // Generated by proc macro.
  type ParentState: States;
}

#[derive(States)]
#[state(parent = A)] // Optional.
enum B { ... }
```

And then in new `apply_state_transition_systems` we need to add more constraints for the state relative to parent sets. For example,

```
run_enter_schedule::<S>.in_set(StateTransitionSet::OnEnter)
```

should become something like

```
run_enter_schedule::<S>
  .in_set(StateTransitionSet::OnEnter)
  .after(run_enter_schedule::<S::ParentState>) // this should add dependency only is S != S::ParentState
```

This would just set state transitions correctly ordered relative to parent.

Now we have ordering of enter/exit.

The tricky part is automatically change parent state and child states when current state changes. This would be harder, I have some ideas how to do it, but it would be a wall of text. But the bottom line is:

> this doesn't feel like it gets us any closer to that

It gets us a little closer to that:
- It adds some constraints which must be preserved when substates implemented
- It adds a test which must still pass if states are hierarchical

**Comment by maniwani** [comment]
> I'd like a last set of opinions from at least one of you. I'm nervous there's something I'm missing. The loss of atomicity is annoying, but I think that the benefits are worth it.

Is this PR motivated by sub-states? It's not super clear to me what the goal or benefit here is.

> What we really need is a solution for sub-states and this doesn't feel like it gets us any closer to that.
> [Sub-states] have an inside-outside relationship with each other. The ordering described in this PR is just arbitrary. 

I tend to agree. I would prefer to handle that with observers (e.g. #10839). The way observers interact with command application order would give us the stack/recursion we need to make nested state transitions happen in the desired order.

I'm ambivalent about the current convention of running all transitions in a single place in `Main`. If we could adopt a convention where parent-child is the only relationship we acknowledge between states, then (with observers) there would be no issue with state transitions taking effect "immediately" (next sync point).

**Comment by stepancheg** [comment]
> Is this PR motivated by sub-states?

Yes, I use substates in my app (but it is complex and not scalable implementation).

<details>
<summary>For the record, this is what I do now.</summary>

There are several states in my "game". Let's call them:
- we are building a house
  - we are picking a place to build a house
  - we are choosing the house size
- we are hovering over the entities on the map
- various unknown states like "doing nothing" or "console open" or "main menu" or "building something else I didn't fully implement yet"

When known state is entered in one of known states, the application:
- changes cursor
- changes help panel on the screen

When known state is exited, the application:
- resets cursor to default
- clear help panel

Unknown state enter/exit has nothing to do with cursor or panel.

So when state is unknown, we don't leave cursor and help in previous state.

Currently I implemented this logic as: there's a resource like

```
#[derive(Resource, Default)]
struct WantCursor {
  want_cursor: Cursor,
  clear_cursor: bool,
}
```

and there's a system:

```
fn update_cursor_on_request_after_state_changes(
  cursor: ResMut<WantCursor>
) {
  if cursor.is_changed() {
    if cursor.want_cursor {
      // actually set cursor
    } else {
      // reset cursor to default
    }
    *cursor = WantCursor::default();
  }
}
```

So `OnEnter` of known state cursor resource `want_cursor` is set. And `OnExit` `clear_cursor` is set. So if both `OnEnter` and `OnExit` run, both `want_cursor` and `clear_cursor` are set, but `want_cursor` has precedence.

If all `OnExit` were executed before all `OnEnter`, I would not need this `clear_cursor`/extra system hacks, and could set/clear the cursor right in `OnEnter`/`OnExit`.
</details>

> I'm ambivalent about our convention of running all transitions in a single place in `Main`

I actually think this is drawback of current design. Instead we could schedule all the systems which depend on the state after all the `OnEnter`/`OnExit` systems — probably even in `Update` schedule — to increase parallelism.

I raised this issue numerous times (last time, [here](https://github.com/bevyengine/bevy/discussions/10851)) and was not understood.

> I would prefer to handle that with observers (e.g. https://github.com/bevyengine/bevy/pull/10839)

I read this PR.
- It is about entities, not resources
- PR description does not say anything about order
- Observers seem to access the world, so no parallelism? Seems to be drawback.
- Overall, seems not very related to this issue. But maybe I probably did not understand it.

**Comment by maniwani** [comment]
> - It is about entities, not resources.

No, it's not. The PR gives examples with structural change events like `OnAdd`, but users can define their own events to emit (look at the `Resize` example), and just know that specifying a `target` is optional. A user could emit a `Resize` event while referring to no entity in particular and register observers to listen for that.

Being able to register callbacks on internal ECS changes is just one of the benefits of observers, not their sole purpose.

> - PR description does not say anything about order.

I have insider knowledge. Just take my word for it. If commands `A` and `B` are queued and `A` runs and does something that triggers observers, any commands those produce (and so on, "recursively") will run before `B`.

> - Observers seem to access the world, so no parallelism?

Correct.[^1]

> Seems to be drawback.

The status quo is that different states do not transition in parallel. They *can't* transition in parallel because those are distinct schedules (dependency graphs), and the reasons why *that* is the case are: 

- State transitions can make structural changes. There were no reasons to consider letting multiple transitions overlap. (No clue if we think differently now, but definitely not back then.)
- States are allowed to transition in any schedule and even at multiple points in the same schedule.[^2] A key conclusion of the scheduling V3 work was that the dependency graph can't represent cycles (obvious, I know) or branches. Doing flow control requires exclusive access (i.e. you can only loop/branch *inside* a command or an exclusive system).
  
Maybe there's an alternative where we say there's a singular, all-encompassing "app state" with two singular "enter" and "exit" schedules owning the systems from all user states (with `OnEnter(S)` and `OnExit(S)` as sets) that we run for *any* transition. idk

[^1]: I think the community values "running systems in parallel" too highly. Parallel system execution has pretty significant overhead and we lack really solid evidence that it handily wins against the more basic, serial execution.

[^2]: This is why I brought up observers and also why I'm ambivalent about this unique property of the `Main` schedule. `Main` having one spot where transitions are supposed to happen was an executive decision.

**Comment by stepancheg** [comment]
> They can't transition in parallel because those are distinct schedules

I would actually prefer to change that. Instead of API

```
app.add_systems(OnEnter(X), enter_x)
```

I would have API

```
app.add_on_enter(X, enter_x)
```

so Bevy would have more freedom to change state internals. In particular, all the systems will be arranged relative to each other just with system sets.

> I think the community values "running systems in parallel" too highly

It is an interesting topic.

I think this is not what community "values", but this is what community "expects".

For example, if execution is multi-threaded
- I make sure a system does not grab more than it needs to (to allow other heavy systems run concurrently with it)
- I'm careful about not adding unnecessary dependencies to the systems
- at the same time I avoid optimizations of heavy systems, because machine is busy running other heavy systems anyway

If Bevy main schedule was single threaded, I could live with it, but it would significantly change my coding style, for example:
- I would not worry about grabbing too many resources mutably
  - in particular, I would probably add a param like `EasyParamEverythingMyGameEverMayNeed`
- generally I would make systems larger to simplicity of coding (because it does not affect main schedule execution time)
- at the same time I would start parallelizing work inside the systems, because each system is now now critical path

Basically, single-threaded vs multithreaded imply very different coding styles. Either is probably fine.

If Bevy advertises multi-threaded setup, it needs to make sure it utilizes this multi-threading as much as possible. In particular, in state transitions and things such observers.

Maybe Bevy should stop doing that.

**Comment by stepancheg** [comment]
I have rewritten my app with custom state implementation, so I don't need this PR as much as before.

Closing to reduce the queue size. If something changes their mind and thinks it would be good idea to merge it, please ping me, and I'll reopen it and rebase.

**Comment by rdrpenguin04** [comment]
@stepancheg This seems like an important PR for sake of consistency if nothing else; it helps the graph make more sense.
